### PR TITLE
CSB-34 Move tokenization of the dataset out of training loop to prevent redundant tokenizing after every epoch

### PR DESCRIPTION
**Pull request by IllBeWatchingU**

### CODE REVIEW

File: model_utilities/train/finetuner.py [hunk 1]
```diff
@@ -42,53 +44,68 @@ def __plot_losses(self, training_losses: list[float], validation_losses: list[fl
         plt.ylabel("Loss")
         plt.legend()
         plt.savefig("./debug/graphs/losses.png")
-
-    def __train(self, model: AutoModelForSeq2SeqLM, tokenizer: NllbTokenizer, dataset: ds.Dataset, optimizer: Adafactor,
+        
+    def __tokenize(self, dataset: ds.Dataset, src_lang: str, tgt_lang: str, tokenizer: NllbTokenizerFast, max_length : int) -> BatchEncoding:
+        tokenizer.src_lang = src_lang
+        tokenizer.tgt_lang = tgt_lang
+        tokenized_input = tokenizer(dataset[src_lang], text_target=dataset[tgt_lang], return_tensors='pt', padding=True, truncation=True, max_length=max_length)
```
Comment (filip-szweda) [inline model_utilities/train/finetuner.py:51]:
```suggestion
        tokenized_input = tokenizer(dataset[src_lang], text_target=dataset[tgt_lang], return_tensors="pt", padding=True, truncation=True, max_length=max_length)
```
Let's keep the uniform quoting style across the codebase.

File: model_utilities/train/finetuner.py [hunk 2]
```diff
@@ -42,53 +44,68 @@ def __plot_losses(self, training_losses: list[float], validation_losses: list[fl
...
+        tokenized_input.labels[tokenized_input.labels == tokenizer.pad_token_id] = -100
```
Comment (filip-szweda) [inline model_utilities/train/finetuner.py:52]: Why -100 exactly? Please either move it to a variable and then use it, or explain its meaning in a comment.

File: model_utilities/train/finetuner.py [hunk 3]
```diff
@@ -42,53 +44,68 @@ def __plot_losses(self, training_losses: list[float], validation_losses: list[fl
...
+        return tokenized_input
+    
+    def __tokenize_and_cleanup_dataset_pairs(self, dataset: ds.Dataset, tokenizer: NllbTokenizerFast, max_length : int) -> tuple[Dataset, Dataset]:
```
Comment (filip-szweda) [inline model_utilities/train/finetuner.py:55]: This function violates the single responsibility principle. Please refactor by creating a `__normalize_dataset` function that takes `tokenized_dataset`, `tokenizer`, and `max_length` as parameters. Create `tokenized_train_dataset_pl_to_csb` and `tokenized_train_dataset_csb_to_pl` outside of this function, then pass them as arguments to `__normalize_dataset`.

File: model_utilities/train/finetuner.py [hunk 4]
```diff
@@ -42,53 +44,68 @@ def __plot_losses(self, training_losses: list[float], validation_losses: list[fl
...
+        tokenized_train_dataset_pl_to_csb = dataset.map(
+            lambda dataset: self.__tokenize(dataset, "pol_Latn", "csb_Latn", tokenizer, max_length),
```
Comment (filip-szweda) [inline model_utilities/train/finetuner.py:57]: Save "pol_Latn" and "csb_Latn" as `ModelFinetuner` attributes, as we use these values multiple times through method calls of the class.

File: model_utilities/train/finetuner.py [hunk 5]
```diff
@@ -43,52 +44,83 @@ def __plot_losses(self, training_losses: list[float], validation_losses: list[fl
         plt.legend()
         plt.savefig("./debug/graphs/losses.png")
 
-    def __train(self, model: AutoModelForSeq2SeqLM, tokenizer: NllbTokenizer, dataset: ds.Dataset, optimizer: Adafactor,
+    def __tokenize_dataset(self, dataset: ds.Dataset, tokenizer: NllbTokenizerFast, max_length: int) -> Dataset:
+        # Not meant to be called outside of this function
+        def tokenize_fn(dataset: ds.Dataset, src_lang: str, tgt_lang: str, tokenizer: NllbTokenizerFast, max_length: int) -> BatchEncoding:
```
Comment (filip-szweda) [inline model_utilities/train/finetuner.py:49]:
```suggestion
        def tokenize(dataset: ds.Dataset, src_lang: str, tgt_lang: str, tokenizer: NllbTokenizerFast, max_length: int) -> BatchEncoding:
```
It's unnecessary to label this function as a function.

File: model_utilities/train/finetuner.py [hunk 6]
```diff
@@ -43,52 +44,83 @@ def __plot_losses(self, training_losses: list[float], validation_losses: list[fl
...
+            tokenizer.tgt_lang = tgt_lang
+            tokenized_input = tokenizer(dataset[src_lang], text_target=dataset[tgt_lang], return_tensors='pt', padding=True, truncation=True, max_length=max_length)
+
+            # -100 is the default value ignored by pytorch
+            # TODO: Check performance after removing this line
+            tokenized_input.labels[tokenized_input.labels == tokenizer.pad_token_id] = -100
+            return tokenized_input
+
+        tokenized_train_dataset_pl_to_csb = dataset.map(
+            lambda row: tokenize_fn(row, "pol_Latn", "csb_Latn", tokenizer, max_length),
```
Comment (filip-szweda) [inline model_utilities/train/finetuner.py:60]: See: https://github.com/kashubian-translator/pl-csb-model/pull/20#discussion_r1855259203

File: model_utilities/train/finetuner.py [hunk 7]
```diff
@@ -43,52 +44,83 @@ def __plot_losses(self, training_losses: list[float], validation_losses: list[fl
...
+            batched=True
+        )
+        tokenized_train_dataset_csb_to_pl = dataset.map(
+            lambda row: tokenize_fn(row, "csb_Latn", "pol_Latn", tokenizer, max_length),
+            batched=True
+        )
+
+        tokenized_dataset = ds.interleave_datasets([tokenized_train_dataset_pl_to_csb, tokenized_train_dataset_csb_to_pl])
```
Comment (filip-szweda) [inline model_utilities/train/finetuner.py:68]: Why do we interleave the datatests?

File: model_utilities/train/finetuner.py [hunk 8]
```diff
@@ -43,52 +44,83 @@ def __plot_losses(self, training_losses: list[float], validation_losses: list[fl
...
+        tokenized_dataset = tokenized_dataset.remove_columns(["pol_Latn", "csb_Latn"])
+
+        return tokenized_dataset
+
+    def __train(self, model: AutoModelForSeq2SeqLM, tokenizer: NllbTokenizerFast, dataset: ds.Dataset, optimizer: Adafactor,
                 config: ConfigParser) -> None:
         self.__log_train_config(config)
 
+        # Prepare variables
```
Comment (filip-szweda) [inline model_utilities/train/finetuner.py:77]: This is an unnecessary comment. Perhaps a log would be better to indicate the process?

File: model_utilities/train/finetuner.py [hunk 9]
```diff
@@ -43,52 +44,83 @@ def __plot_losses(self, training_losses: list[float], validation_losses: list[fl
...
+        best_val_loss = float("inf")
+        patience_counter = 0
 
         training_losses = []
         validation_losses = []
         scheduler = get_constant_schedule_with_warmup(optimizer, num_warmup_steps=int(train_conf["warmup_steps"]))
 
-        train_batches = dataset["train"].batch(batch_size=int(train_conf["batch_size"]))
-        val_batches = dataset["validation"].batch(batch_size=int(train_conf["batch_size"]))
+        # Prepare dataset
```
Comment (filip-szweda) [inline model_utilities/train/finetuner.py:91]: Same as with `# Prepare variables`.

File: model_utilities/train/finetuner.py [hunk 10]
```diff
@@ -43,52 +44,83 @@ def __plot_losses(self, training_losses: list[float], validation_losses: list[fl
         plt.legend()
         plt.savefig("./debug/graphs/losses.png")
 
-    def __train(self, model: AutoModelForSeq2SeqLM, tokenizer: NllbTokenizer, dataset: ds.Dataset, optimizer: Adafactor,
+    def __tokenize_dataset(self, dataset: ds.Dataset, tokenizer: NllbTokenizerFast, max_length: int) -> Dataset:
+        # Not meant to be called outside of this function
```
Comment (filip-szweda) [inline model_utilities/train/finetuner.py:48]: This is an unnecessary comment. `tokenize` inside `__tokenize_dataset` clearly indicates that.

File: model_utilities/train/finetuner.py [hunk 11]
```diff
@@ -43,52 +44,83 @@ def __plot_losses(self, training_losses: list[float], validation_losses: list[fl
         plt.legend()
         plt.savefig("./debug/graphs/losses.png")
 
-    def __train(self, model: AutoModelForSeq2SeqLM, tokenizer: NllbTokenizer, dataset: ds.Dataset, optimizer: Adafactor,
+    def __tokenize_dataset(self, dataset: ds.Dataset, tokenizer: NllbTokenizerFast, max_length: int) -> Dataset:
```
Comment (filip-szweda) [inline model_utilities/train/finetuner.py:47]: Please add a detailed, step-by-step explanation within the comments to clarify why we are tokenizing the dataset using this specific, undocumented approach. Understanding this process solely from the code is challenging without context about the data being handled.

File: model_utilities/train/finetuner.py [hunk 12]
```diff
@@ -43,52 +44,83 @@ def __plot_losses(self, training_losses: list[float], validation_losses: list[fl
...
             epoch_training_losses = []
-            for i in range(batch_num):
-                batch = train_batches[i]
-                # Swap the direction of translation for some batches
-                batch = list(batch.items())
-                random.shuffle(batch)
-
-                (lang1, xx), (lang2, yy) = batch
+            for batch in train_dataloader:
+                batch = {k: v.to(model.device) for k, v in batch.items()}
```
Comment (filip-szweda) [inline model_utilities/train/finetuner.py:121]: What are `k` and `v` other than a key and a value?

File: model_utilities/train/finetuner.py [hunk 13]
```diff
@@ -103,31 +135,26 @@ def __train(self, model: AutoModelForSeq2SeqLM, tokenizer: NllbTokenizer, datase
...
             training_losses.append(avg_training_loss)
 
             epoch_validation_losses = []
             with torch.no_grad():
-                for batch in val_batches:
-                    batch = list(batch.items())
-                    (lang1, xx), (lang2, yy) = batch
+                self.__logger.info(f"Epoch {epoch + 1}/{num_epochs}: Starting validation")
+                val_progress_bar = trange(val_batch_num * batch_size)
+                for batch in validation_dataloader:
```
Comment (filip-szweda) [inline model_utilities/train/finetuner.py:147]: This loss calculation code is used two times, for training and for validation. Please move it to a function.

File: model_utilities/train/finetuner.py (hunk 5 above)
Comment (jan-Kulpas) [reply model_utilities/train/finetuner.py:49]: personally i think it's useful because it implies it being passed as parameter to the map() function

File: model_utilities/train/finetuner.py (hunk 7 above)
Comment (jan-Kulpas) [reply model_utilities/train/finetuner.py:68]: The previous approach was to tokenize after we randomize the direction in which we are translating (literally pick a pair and then pass one as input and one as labels to the model at random)

Because we moved tokenization outside the training loop we thought it would be better to just train the model in both directions for each sentence. It was easier to merge the duplicated dataset to preserve the inputs being in the input_ids column and labels in the labels column than to swap it around during the training loop.

We could concatenate the datasets instead of interleaving them but I thought it's just nicer this way and maybe more closely represents the previous implementation.

No idea if it will yield better results than concat though. We could try both?

File: model_utilities/train/finetuner.py (hunk 8 above)
Comment (jan-Kulpas) [reply model_utilities/train/finetuner.py:77]: eh, maybe. i like it though as it provides a visible break in code between the steps.

I agree though that it looks dumb because I stopped adding them midway. will consider logging.

File: model_utilities/train/finetuner.py (hunk 12 above)
Comment (jan-Kulpas) [reply model_utilities/train/finetuner.py:121]: This line is copied directly from HF documentation. Each row in the batch is a dict and we have to call `to(device)` to allow it to be loaded by cuda.

This was also done before but per row rather than per batch

File: model_utilities/train/finetuner.py (hunk 13 above)
Comment (jan-Kulpas) [reply model_utilities/train/finetuner.py:147]: im scared

moving this to a different story

**Review by filip-szweda — APPROVED** [review]
